infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -192,15 +193,49 @@ private PollReport pollRound(
 
         tunePollReportTimeouts(router, failedNodesAggregated, allConnectedNodes);
 
-        PollReport latestReport = reports.get(reports.size() - 1);
-
         return PollReport.builder()
                 .pollEpoch(epoch)
                 .connectedNodes(ImmutableSet.copyOf(connectedNodesAggregated))
                 .failedNodes(ImmutableSet.copyOf(failedNodesAggregated))
                 .wrongEpochs(ImmutableMap.copyOf(wrongEpochsAggregated))
                 .currentLayoutSlotUnFilled(isCurrentLayoutSlotUnFilled(layout, wrongEpochsAggregated.keySet()))
-                .clusterState(latestReport.getClusterState())
+                .clusterState(getAggregatedState(reports))
+                .build();
+    }
+
+    /**
+     * Aggregating pollRound reports and provide aggregated view of the cluster.
+     *
+     * @param reports pollRound reports
+     * @return cluster state
+     */
+    private ClusterState getAggregatedState(List<PollReport> reports) {
+        Map<String, NodeState> stateMap = new HashMap<>();
+
+        reports.forEach(report -> {
+            ClusterState currState = report.getClusterState();
+            for (NodeState nodeState : currState.getNodes().values()) {
+                String endpoint = nodeState.getConnectivity().getEndpoint();
+                NodeConnectivityType connType = nodeState.getConnectivity().getType();
+
+                if (!stateMap.containsKey(endpoint)) {
+                    stateMap.put(endpoint, nodeState);
+                    continue;
+                }
+
+                //Don't replace by a new state if the new state is not CONNECTED
+                if (stateMap.get(endpoint).getConnectivity().getType() == NodeConnectivityType.CONNECTED
```

Comment:
This can just be: ```if (stateMap.get(endpoint).getConnectivity().getType() == NodeConnectivityType.CONNECTED)``` right?

Reply:
no
you can replace CONNECTED node state only by another connected node state

Reply:
Ok, that was my point. Why replace when you can even leave the first one.
But I see you would want to keep the latest node state.